HADOOP-13327 Output Stream Specification

steveloughran opened the pull request:
HADOOP-13327 Output Stream Specification

Change-Id: I1b6bc258a40a8bd57879d9edc3e5bb1303f0fff2

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 32 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 22 | Maven dependency ordering for branch |
| +1 | mvninstall | 1144 | trunk passed |
| +1 | compile | 1118 | trunk passed |
| +1 | checkstyle | 195 | trunk passed |
| +1 | mvnsite | 221 | trunk passed |
| +1 | shadedclient | 1097 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 348 | trunk passed |
| +1 | javadoc | 177 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 25 | Maven dependency ordering for patch |
| +1 | mvninstall | 185 | the patch passed |
| +1 | compile | 1026 | the patch passed |
| -1 | javac | 1026 | root generated 2 new + 1483 unchanged - 0 fixed = 1485 total (was 1483) |
| -0 | checkstyle | 193 | root: The patch generated 4 new + 32 unchanged - 3 fixed = 36 total (was 35) |
| +1 | mvnsite | 235 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 2 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 647 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 330 | the patch passed |
| +1 | javadoc | 147 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 529 | hadoop-common in the patch passed. |
| -1 | unit | 7899 | hadoop-hdfs in the patch failed. |
| +1 | unit | 281 | hadoop-aws in the patch passed. |
| +1 | unit | 70 | hadoop-azure-datalake in the patch passed. |
| +1 | asflicense | 50 | The patch does not generate ASF License warnings. |
| | | 15820 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.protocol.datatransfer.sasl.TestSaslDataTransfer |
|   | hadoop.hdfs.TestAclsEndToEnd |
|   | hadoop.TestRefreshCallQueue |
|   | hadoop.hdfs.TestDFSStripedOutputStream |
|   | hadoop.hdfs.server.blockmanagement.TestSequentialBlockGroupId |
|   | hadoop.hdfs.server.mover.TestMover |
|   | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.hdfs.TestUnsetAndChangeDirectoryEcPolicy |
|   | hadoop.hdfs.client.impl.TestBlockReaderLocal |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithFailureWithRandomECPolicy |
|   | hadoop.hdfs.TestErasureCodingPolicies |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithFailure |
|   | hadoop.hdfs.TestErasureCodingPoliciesWithRandomECPolicy |
|   | hadoop.hdfs.server.namenode.TestAuditLogs |
|   | hadoop.hdfs.TestFileAppend2 |
|   | hadoop.metrics2.sink.TestRollingFileSystemSinkWithHdfs |
|   | hadoop.hdfs.TestDFSClientExcludedNodes |
|   | hadoop.hdfs.TestStripedFileAppend |
|   | hadoop.hdfs.TestReadStripedFileWithMissingBlocks |
|   | hadoop.hdfs.TestGetFileChecksum |
|   | hadoop.hdfs.TestEncryptedTransfer |
|   | hadoop.hdfs.TestDFSStripedInputStream |
|   | hadoop.hdfs.TestReconstructStripedFile |
|   | hadoop.hdfs.TestErasureCodingPolicyWithSnapshotWithRandomECPolicy |
|   | hadoop.hdfs.TestDatanodeDeath |
|   | hadoop.fs.TestSymlinkHdfsFileSystem |
|   | hadoop.hdfs.TestSetrepIncreasing |
|   | hadoop.hdfs.tools.offlineEditsViewer.TestOfflineEditsViewer |
|   | hadoop.hdfs.TestReadStripedFileWithDecoding |
|   | hadoop.hdfs.TestDecommission |
|   | hadoop.metrics2.sink.TestRollingFileSystemSinkWithSecureHdfs |
|   | hadoop.hdfs.TestReadStripedFileWithDNFailure |
|   | hadoop.hdfs.TestDecommissionWithStriped |
|   | hadoop.hdfs.server.namenode.TestBlockPlacementPolicyRackFaultTolerant |
|   | hadoop.hdfs.server.datanode.TestBPOfferService |
|   | hadoop.hdfs.server.balancer.TestBalancer |
|   | hadoop.hdfs.TestLeaseRecovery2 |
|   | hadoop.hdfs.TestFileCorruption |
|   | hadoop.hdfs.server.namenode.TestPersistentStoragePolicySatisfier |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-575/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  xml  |
| uname | Linux 4ab62737f35d 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / fb851c9 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/1/artifact/out/diff-compile-javac-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/1/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/1/testReport/ |
| Max. process+thread count | 3303 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-aws hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

steveloughran commented:
HDFS test have run out of memory

```

[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 64.999 s <<< FAILURE! - in org.apache.hadoop.TestRefreshCallQueue
[ERROR] testRefreshCallQueueWithFairCallQueue(org.apache.hadoop.TestRefreshCallQueue)  Time elapsed: 4.751 s  <<< ERROR!
java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:717)
	at org.apache.hadoop.fs.FileSystem$Statistics.<clinit>(FileSystem.java:3707)
	at org.apache.hadoop.fs.FileSystem.getStatistics(FileSystem.java:4175)
	at org.apache.hadoop.fs.FileSystem.initialize(FileSystem.java:295)
	at org.apache.hadoop.fs.RawLocalFileSystem.initialize(RawLocalFileSystem.java:99)
	at org.apache.hadoop.fs.LocalFileSystem.initialize(LocalFileSystem.java:47)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:3324)
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:136)
	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:3373)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:3341)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:491)
	at org.apache.hadoop.fs.FileSystem.getLocal(FileSystem.java:447)
	at org.apache.hadoop.hdfs.server.datanode.checker.StorageLocationChecker.check(StorageLocationChecker.java:159)
	at org.apache.hadoop.hdfs.server.datanode.DataNode.makeInstance(DataNode.java:2785)
	at org.apache.hadoop.hdfs.server.datanode.DataNode.instantiateDataNode(DataNode.java:2698)
	at org.apache.hadoop.hdfs.MiniDFSCluster.startDataNodes(MiniDFSCluster.java:1697)
	at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:913)
```

steveloughran commented:
I consider that to be an HDFS problem: not mine. This test is ready for review

bgaborg commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCreateTest.java:
```diff
@@ -332,4 +336,78 @@ public void testCreateMakesParentDirs() throws Throwable {
+      assertEquals("hflush support", supportsFlush, doesHFlush);
+      boolean doesHSync = out.hasCapability(
+          StreamCapabilities.StreamCapability.HSYNC.getValue());
+      assertEquals("HFlush support", supportsSync, doesHSync);
```
> "HFlush support"

This is hsync support, not hflush

busbey commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Syncable.java:
```diff
@@ -27,16 +27,16 @@
    * this call, new readers will see the data.
    * @throws IOException if any error occurs
    */
-  public void hflush() throws IOException;
```
he didn't. it's an interface so it's implicitly public.

busbey commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Syncable.java:
```diff
@@ -27,16 +27,16 @@
    * all the way to the disk device (but the disk may have it in its cache).
    * @throws IOException if error occurs
    */
-  public void hsync() throws IOException;
```
he didn't. it's an interface so it's implicitly public.

ehiggs commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+#### Preconditions
+
+```python
+hasCapability(Stream. "hflush")
````
```suggestion
hasCapability(Stream, "hflush")
```

ehiggs commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+read.
+1. Use `getFileStatus(FS, Path)` to query the updated length of the file as
+recorded in the metadata.
+1. If `Status.length &gt pos`, the file has grown.
```
```suggestion
1. If `Status.length &gt; pos`, the file has grown.
```

steveloughran replied:
fixed

ehiggs commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+to the stream.
+
+For anyone thinking "this is a violation of this specification" —they are correct.
+The local filesystem is intended for testing, rather than production use.
```
>The local filesystem is intended for testing, rather than production use.

Not sure that's entirely true. People try to use Hadoop on HPC systems where the parallel file system is LocalFileSystem. There is a [`gpfs://`](https://www.ibm.com/support/knowledgecenter/en/STXKQY_4.1.1/com.ibm.spectrum.scale.v4r11.adv.doc/bl1adv_hadoop.htm) for Spectrum Scale, but I think, e.g. Lustre, Panasas, etc have to make due with `file://`.

steveloughran replied:
yeah, says "was". FWIW I think we could do more here

ehiggs commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+1. If a file was created with `overwrite = True`, the existing data my still
+be visible: `data(FS', path) = data(FS, path)`.
+
+1. The check for existing data in a `create()` call with `overwrite=False`, may
```
To be consistent with the `overwrite==` above in the document...

```suggestion
1. The check for existing data in a `create()` call with `overwrite==false`, may
```

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/CanSetDropBehind.java:
```diff
@@ -36,6 +36,6 @@
    *                                        setting the drop-behind.
    */
-  public void setDropBehind(Boolean dropCache) 
+  void setDropBehind(Boolean dropCache)
```
IDE & checkstyle was complaining, so while I was here....

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StoreImplementationUtils.java:
```diff
@@ -0,0 +1,81 @@
+   * Probe for an object having a capability; returns true
+   * iff the stream implements {@link StreamCapabilities} and its
+   * {@code hasCapabilities()} method returns true for the capability.
+   * This is a private method intended to provided a common implementation
```
Sorry, its an `@Private` attr for anything in hadoop-*. I'll clarify that.

steveloughran replied:
I'm confused and wrong; you are right. I'll make it package-private

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,205 @@
+   * However, to guarantee the state is constant through the life of an
+   * operation, updates must be through the synchronized methods.
+   */
+  private volatile State state = State.Open;
```
stops reads blocking on what should be a low-cost lookup. But yes, even though I think *I* know what I'm doing, I don't know what the compiler and CPU are doing, or what whoever reads it next will do. I think I started off with a non-volatile, but decided to make it volatile for consistency. Even if the getter directly/indirectly goes to synchronized, its potentially out of date as soon as the value is used outside the synchronized block, so I didn't see much merit in going overboard in locks here

busbey replied:
The Atomic* types in Java don't involve locking; they're essentially just API wrappers around volatile to make it easier to reason about for folks who aren't familiar with the Java Memory Model.

Details:

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/package-summary.html#package.description

steveloughran replied:
moving to volatile. Makes compare-and-set someone else's problem

steveloughran replied:
removed the class

ajfabbri commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,205 @@
+   * Change state to error and stores first error so it can be re-thrown.
+   * If already in error: return previous exception.
+   * @param ex the exception to record
+   * @return the exception set when the error state was entered.
```
Hi Sean! Current code, yes, but would we ever actually need that information? (Maybe in a unit test). It is going to be in error when the call returns either way.

ajfabbri commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractFSContractTestBase.java:
```diff
@@ -272,7 +272,7 @@ protected void handleRelaxedException(String action,
       throw e;
     }
-    LOG.warn("The expected exception {}  was not the exception class" +
+    LOG.warn("The expected exception {} was not the exception class" +
```
IIRC an extra last parameter of type Throwable is allowed for Logger

ehiggs replied:
```java
    // org.slf4j.Logger
    public void warn(String msg);
    public void warn(String format, Object arg);
    public void warn(String format, Object... arguments); // <-- I think it will choose this. Maybe in the impl it checks the type to see if it's throwable? 🤷‍♂️ 
    public void warn(String format, Object arg1, Object arg2);
    public void warn(String msg, Throwable t);
    public void warn(Marker marker, String msg);
    public void warn(Marker marker, String format, Object arg);
    public void warn(Marker marker, String format, Object arg1, Object arg2);
    public void warn(Marker marker, String format, Object... arguments);
    public void warn(Marker marker, String msg, Throwable t);
```

steveloughran replied:
throwables come at the end, they don't go into the error text

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 25 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for branch |
| +1 | mvninstall | 966 | trunk passed |
| +1 | compile | 991 | trunk passed |
| +1 | checkstyle | 199 | trunk passed |
| +1 | mvnsite | 285 | trunk passed |
| +1 | shadedclient | 1164 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 295 | trunk passed |
| +1 | javadoc | 172 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 22 | Maven dependency ordering for patch |
| +1 | mvninstall | 159 | the patch passed |
| +1 | compile | 880 | the patch passed |
| -1 | javac | 880 | root generated 2 new + 1483 unchanged - 0 fixed = 1485 total (was 1483) |
| -0 | checkstyle | 233 | root: The patch generated 4 new + 32 unchanged - 3 fixed = 36 total (was 35) |
| +1 | mvnsite | 243 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 3 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 674 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 354 | the patch passed |
| +1 | javadoc | 154 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 509 | hadoop-common in the patch passed. |
| -1 | unit | 4737 | hadoop-hdfs in the patch failed. |
| +1 | unit | 292 | hadoop-aws in the patch passed. |
| +1 | unit | 66 | hadoop-azure-datalake in the patch passed. |
| +1 | asflicense | 55 | The patch does not generate ASF License warnings. |
| | | 12300 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.hdfs.server.datanode.TestDataNodeHotSwapVolumes |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-575/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  xml  |
| uname | Linux faa498435078 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 73f7b04 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/2/artifact/out/diff-compile-javac-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/2/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/2/testReport/ |
| Max. process+thread count | 4987 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-aws hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,205 @@
+ *   Error: all operations will raise the previously recorded exception.
+ *   Closed: operations will be rejected.
+ * </pre>
+ */
```
This is based on work from the Azure side of things; stuff which wasn't being done right elsewhere, as it turned out. Will clarify.

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+-->
+
+<!-- MACRO{toc|fromDepth=1|toDepth=3} -->
+
```
This is the same specification syntax as the rest of the filesystem specification. I'll clarify that in the intro

steveloughran replied:
done

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,205 @@
+   * Check a stream is open.
+   * If in an error state: rethrow that exception. If closed,
+   * throw an exception about that.
+   * @throws IOException if the stream is not open.
```
done

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,205 @@
+   * @return return true iff the steam was in the state at the time
+   * of checking.
+   */
+  public synchronized boolean isInState(State expected) {
```
none, I'll remove it for you.

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+`org.apache.hadoop.hdfs.DFSOutputStream`.
+* The subclass `org.apache.hadoop.hdfs.DFSStripedOutputStream` supports erasure
+coding: it removes the `Syncable` behaviors from the base class.
+* The output streams `org.apache.hadoop.fs.FSOutputSummer` and
```
Cut them out; will keep maintenance down

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+buffer: List[byte]
+```
+
+The destination path of the stream, `path` can be tracked to form a triple
````
fixed. thanks

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+`Path, open, buffer`
+
+
+```python
````
done

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+```
+
+(Immediately) after `Syncable` operations which flush data to the filesystem,
+the data at the stream's destination path must match that of
````
s/must/r/MUST/ for that consistency.

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+`buffer`. That is, the following condition holds:
+
+```python
+FS'.Files(path) == buffer
````
now I'm confused. Can you clarify this a bit. As I'm very much trying to talk about durability of the data.

steveloughran replied:
yes. Actually pulled over the definition fom the later close() call
```
FS' = FS where data(FS', path) == buffer
```

steveloughran commented:
just rebased to trunk; not tested. I see there's comments I've got to address -will do

hadoop-yetus commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,203 @@
+  }
+
+  /**
+   * Create for a path. 
```
whitespace:end of line

hadoop-yetus replied:
whitespace:end of line

hadoop-yetus commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,203 @@
+  /**
+   * Create for a path. 
+   * @param path optional path for exception messages.
+   */  
```
whitespace:end of line

hadoop-yetus replied:
whitespace:end of line

hadoop-yetus commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,203 @@
+/**
+ * Models a stream's state and can be used for checking this state before
+ * any operation.
+ * 
```
whitespace:end of line

hadoop-yetus replied:
whitespace:end of line

hadoop-yetus commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,203 @@
+  private IOException exception;
+
+  /**
+   * Create for a path. 
```
whitespace:end of line

hadoop-yetus replied:
whitespace:end of line

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 33 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 64 | Maven dependency ordering for branch |
| +1 | mvninstall | 1135 | trunk passed |
| +1 | compile | 1013 | trunk passed |
| +1 | checkstyle | 122 | trunk passed |
| +1 | mvnsite | 252 | trunk passed |
| +1 | shadedclient | 1039 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 192 | trunk passed |
| 0 | spotbugs | 48 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 442 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for patch |
| +1 | mvninstall | 192 | the patch passed |
| +1 | compile | 960 | the patch passed |
| -1 | javac | 960 | root generated 2 new + 1475 unchanged - 0 fixed = 1477 total (was 1475) |
| -0 | checkstyle | 137 | root: The patch generated 4 new + 78 unchanged - 3 fixed = 82 total (was 81) |
| +1 | mvnsite | 284 | the patch passed |
| -1 | whitespace | 0 | The patch has 4 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 3 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 690 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 241 | the patch passed |
| +1 | findbugs | 513 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 510 | hadoop-common in the patch passed. |
| -1 | unit | 4821 | hadoop-hdfs in the patch failed. |
| +1 | unit | 297 | hadoop-aws in the patch passed. |
| +1 | unit | 95 | hadoop-azure in the patch passed. |
| +1 | unit | 59 | hadoop-azure-datalake in the patch passed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 13109 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.hdfs.tools.TestWebHDFSStoragePolicyCommands |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithFailureWithRandomECPolicy |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithFailure |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.TestFileChecksumCompositeCrc |
|   | hadoop.hdfs.TestReadStripedFileWithMissingBlocks |
|   | hadoop.hdfs.server.datanode.TestDataNodeVolumeFailureReporting |
|   | hadoop.hdfs.tools.TestDFSAdmin |
|   | hadoop.hdfs.tools.TestDFSHAAdminMiniCluster |
|   | hadoop.hdfs.server.datanode.TestDataNodeLifeline |
|   | hadoop.hdfs.TestReconstructStripedFile |
|   | hadoop.hdfs.TestErasureCodingPolicyWithSnapshotWithRandomECPolicy |
|   | hadoop.hdfs.tools.offlineEditsViewer.TestOfflineEditsViewer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-575/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle xml |
| uname | Linux c9f840e2cee6 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 14552d1 |
| Default Java | 1.8.0_212 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/3/artifact/out/diff-compile-javac-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/3/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/3/artifact/out/whitespace-eol.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/3/testReport/ |
| Max. process+thread count | 4729 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-aws hadoop-tools/hadoop-azure hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 75 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for branch |
| +1 | mvninstall | 1095 | trunk passed |
| +1 | compile | 1181 | trunk passed |
| +1 | checkstyle | 140 | trunk passed |
| +1 | mvnsite | 257 | trunk passed |
| +1 | shadedclient | 1068 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 193 | trunk passed |
| 0 | spotbugs | 41 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 435 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for patch |
| +1 | mvninstall | 188 | the patch passed |
| +1 | compile | 1021 | the patch passed |
| -1 | javac | 1021 | root generated 2 new + 1477 unchanged - 0 fixed = 1479 total (was 1477) |
| -0 | checkstyle | 146 | root: The patch generated 4 new + 78 unchanged - 3 fixed = 82 total (was 81) |
| +1 | mvnsite | 280 | the patch passed |
| -1 | whitespace | 0 | The patch has 4 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 3 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 675 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 215 | the patch passed |
| -1 | findbugs | 29 | hadoop-aws in the patch failed. |
| -1 | findbugs | 27 | hadoop-azure in the patch failed. |
| -1 | findbugs | 26 | hadoop-azure-datalake in the patch failed. |
||| _ Other Tests _ |
| +1 | unit | 510 | hadoop-common in the patch passed. |
| -1 | unit | 6478 | hadoop-hdfs in the patch failed. |
| -1 | unit | 32 | hadoop-aws in the patch failed. |
| -1 | unit | 33 | hadoop-azure in the patch failed. |
| -1 | unit | 31 | hadoop-azure-datalake in the patch failed. |
| -1 | asflicense | 47 | The patch generated 1 ASF License warnings. |
| | | 14416 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.8 Server=18.09.8 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle xml |
| uname | Linux 7b3a27325fdb 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / cd967c7 |
| Default Java | 1.8.0_212 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/diff-compile-javac-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/whitespace-eol.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-findbugs-hadoop-tools_hadoop-aws.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-findbugs-hadoop-tools_hadoop-azure.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-findbugs-hadoop-tools_hadoop-azure-datalake.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-unit-hadoop-tools_hadoop-aws.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-unit-hadoop-tools_hadoop-azure.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-unit-hadoop-tools_hadoop-azure-datalake.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 4670 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-aws hadoop-tools/hadoop-azure hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 12 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/5/console |
| versions | git=2.7.4 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 10 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/6/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 14 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/7/console |
| versions | git=2.7.4 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 12 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/8/console |
| versions | git=2.7.4 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 10 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/9/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 13 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/10/console |
| versions | git=2.7.4 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 10 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/11/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 10 | https://github.com/apache/hadoop/pull/575 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/12/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,205 @@
+ *   Closed: operations will be rejected.
+ * </pre>
+ */
+public class StreamStateModel {
```
Everything in this package is declared private and unstable; i'll add it for completeness.

steveloughran replied:
removed the class

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/filesystem.md:
```diff
@@ -598,8 +600,16 @@ The result is `FSDataOutputStream`, which through its operations may generate ne
+the holder of the lock) is not a valid algorithm when working with object stores.
+
+* Object stores may create an empty file as a marker when a file is created.
+However, object stores whith overwrite=true semantics may not implement this atomically,
```
fixed

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+These all return instances of `FSDataOutputStream`, through which data
+can be written through various `write()` methods.
+After a stream's `close()` method is called, all data written to the
+stream MUST BE persisted to the fileysystem and visible to oll other
```
Now you've scared me. Thought it had to be written across >1 replica.

steveloughran replied:
added details on close() in both that method and on an implementor's notes section. Found a JIRA where the Accumulo team "expressed concern":

steveloughran commented:
Had some time to look at the comments here after too long break; more work is needed


## Output stream spec

Needs more review of the comments, especiallys Sean's

## Stream Model + impl.

I now suspect that I need to do a bit more resilience in the failure handling of S3ABlockOutputStream.

Specifically: if 1+ POST has failed in a multipart write, we should make a best effort attempt to abort the upload. Admittedly, the same network or  authentication errors which cause the post to fail are likely to cause the abort operation to fail too -that should not stop us up from at least trying.

Proposed: close() will attempt to abort the upload if a failure occurred earlier.

This is going to be fun to test. Indeed, it's a good argument for WriteOperationHelper to be converted to an interface with a mocking implementation alongside the production one. I might play with some subclassing instead; 


## Plan

I think these two bits of work can be separated; the S3ABlockOutput stream model/locking/aborting is complex enough to merit isolation.

Will follow up with that in a separate PR while retaining this purely for the output stream spec and tests.

This one doesn't apply to trunk no more; once I have gone through the comments I may produce a new PR for the output stream changes too.

hadoop-yetus commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,267 @@
+  private IOException exception;
+
+  /**
+   * Create for a path. 
```
whitespace:end of line

hadoop-yetus commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,267 @@
+  }
+
+  /**
+   * Create for a path. 
```
whitespace:end of line

hadoop-yetus commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,267 @@
+  /**
+   * Create for a path. 
+   * @param path optional path for exception messages.
+   */  
```
whitespace:end of line

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 1465 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 73 | Maven dependency ordering for branch |
| +1 | mvninstall | 1232 | trunk passed |
| +1 | compile | 1070 | trunk passed |
| +1 | checkstyle | 168 | trunk passed |
| +1 | mvnsite | 275 | trunk passed |
| +1 | shadedclient | 1318 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 265 | trunk passed |
| 0 | spotbugs | 43 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 464 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 24 | Maven dependency ordering for patch |
| +1 | mvninstall | 208 | the patch passed |
| +1 | compile | 1129 | the patch passed |
| -1 | javac | 1129 | root generated 2 new + 1890 unchanged - 0 fixed = 1892 total (was 1890) |
| -0 | checkstyle | 172 | root: The patch generated 6 new + 78 unchanged - 3 fixed = 84 total (was 81) |
| +1 | mvnsite | 282 | the patch passed |
| -1 | whitespace | 1 | The patch has 3 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 3 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 826 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 257 | the patch passed |
| +1 | findbugs | 533 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 603 | hadoop-common in the patch failed. |
| -1 | unit | 7305 | hadoop-hdfs in the patch failed. |
| +1 | unit | 126 | hadoop-aws in the patch passed. |
| +1 | unit | 105 | hadoop-azure in the patch passed. |
| +1 | unit | 80 | hadoop-azure-datalake in the patch passed. |
| +1 | asflicense | 65 | The patch does not generate ASF License warnings. |
| | | 17936 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.conf.TestCommonConfigurationFields |
|   | hadoop.hdfs.server.balancer.TestBalancerRPCDelay |
|   | hadoop.hdfs.server.balancer.TestBalancerWithMultipleNameNodes |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/575 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle xml |
| uname | Linux 0d94e2762c82 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ef9d12d |
| Default Java | 1.8.0_222 |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/artifact/out/diff-compile-javac-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/artifact/out/whitespace-eol.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/testReport/ |
| Max. process+thread count | 3318 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs hadoop-tools/hadoop-aws hadoop-tools/hadoop-azure hadoop-tools/hadoop-azure-datalake U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-575/13/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StoreImplementationUtils.java:
```diff
@@ -0,0 +1,81 @@
+   * iff the stream implements {@link StreamCapabilities} and its
+   * {@code hasCapabilities()} method returns true for the capability.
+   * This is a private method intended to provided a common implementation
+   * for input, output streams; only the stronger typed
```
fixed

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+This algorithm works for filesystems which are consistent with metadata and
+data, as well as HDFS. What is important to know is that, in HDFS
+`getFileStatus(FS, path).getLen()==0` does not imply that `data(FS, path)` is
+empty.
```
no way, I'm afraid.

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+to the stream.
+
+For anyone thinking "this is a violation of this specification" —they are correct.
+The local filesystem is intended for testing, rather than production use.
```
changed to "The local filesystem was intended for testing,"

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+Stream2.close()
+```
+
+For anyone wondering why the clients create a 0-byte file in the create call,
````
fixed

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java:
```diff
@@ -344,10 +372,22 @@ private void initMultipartUpload() throws IOException {
+    acquireLock(false);
+    try {
+      synchronized (this) {
+        // this is synchronized to order its execution w.r.t any methods
```
:)

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
```diff
@@ -0,0 +1,857 @@
+  public abstract void write(int b) throws IOException;
+  public void write(byte b[]) throws IOException;
+  public void write(byte b[], int off, int len) throws IOException;
+  public void flush() throws IOException ;
```
fixed

steveloughran commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java:
```diff
@@ -273,14 +268,47 @@ public synchronized void write(int b) throws IOException {
     if (len == 0) {
       return;
     }
+    acquireLock(true);
```
moving this all to a separate patch, eventually

steveloughran commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/StreamStateModel.java:
```diff
@@ -0,0 +1,205 @@
+   * Path; if not empty then a {@link PathIOException} will be raised
+   * containing this path.
+   */
+  private final String path;
```
cut class

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+```
+
+The other postcondition of the operation, `data(FS', path) == []` may not
+hold,
````
added "in which case:"

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+```
+
+Any client reading the data at the path will see the new data.
+The two sync operations, `hflush()` and `hsync()` differ in their durability
````
done

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+Stream' = (path, true, [])
+```
+
+The filesystem `FS'` must contain a 0-byte file at the path:
````
well, in `FS` there was no file

steveloughran replied:
+added the " within a filesystem `FS`" to the create and append references

steveloughran commented on hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md:
````diff
@@ -0,0 +1,857 @@
+data:
+
+```python
+Stream' = (path, true, [])
````
its how it can be modelled after the call. I'll add a ref in filesystem.md

steveloughran commented:
I'm closing this and opening up a peer without any of the S3A changes, #1694. It currently addresses most of the comments of people's reviews, excluding Sean's request about restructuring the model

> Could this be in a section about visibility and not in the model definition? Maybe later. "here's the model, here's how that model works with creation, here's how it works when reading/writing" flows much better and visibility should be in that third part.
